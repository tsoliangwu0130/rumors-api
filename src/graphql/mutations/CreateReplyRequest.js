import {
  GraphQLString,
  GraphQLNonNull,
  GraphQLInt,
  GraphQLObjectType,
  GraphQLEnumType,
} from 'graphql';
import { assertUser } from 'graphql/util';

import client, { processMeta } from 'util/client';

/**
 * @typedef {Object} ReplyRequestParam
 * @property {string} articleId - The article to add reply request to
 * @property {string} userId - The user that submits this request
 * @property {string} appId - The app that the user logged in to
 * @property {string} reason - The reason why the user want to submit this article
 */

/**
 * @param {ReplyRequestParam} params
 * @returns {string} the reply request ID for the givne params
 */
export function getReplyRequestId({ articleId, userId, appId }) {
  return `${articleId}__${userId}__${appId}`;
}

/**
 * @typedef {Object} CreateReplyRequestResult
 * @property {Object} article - The update article instance
 * @property {boolean} isCreated - If the reply request is created
 */

/**
 * Indexes a reply request and increments the replyRequestCount for article
 *
 * @param {ReplyRequestParam} param
 * @returns {CreateReplyRequstResult}
 */
export async function createReplyRequest({
  articleId,
  userId,
  appId,
  reason = '',
}) {
  assertUser({ appId, userId });

  const now = new Date().toISOString();
  const id = getReplyRequestId({ articleId, userId, appId });

  const { result } = await client.update({
    index: 'replyrequests',
    type: 'doc',
    id,
    body: {
      doc: {
        updatedAt: now,
      },
      upsert: {
        articleId,
        userId,
        appId,
        reason,
        feedbacks: [],
        positiveFeedbackCount: 0,
        negativeFeedbackCount: 0,
        createdAt: now,
        updatedAt: now,
      },
    },
  });

  const isCreated = result === 'created';

  const articleUpdateResult = await client.update({
    index: 'articles',
    type: 'doc',
    id: articleId,
    body: {
      script: {
        source: `
          ${isCreated ? 'ctx._source.replyRequestCount += 1;' : ''}
          ctx._source.lastRequestedAt = params.now;
        `,
        params: { now },
      },
    },
    _source: true,
  });

  if (articleUpdateResult.result !== 'updated') {
    throw new Error(`Cannot update article ${articleId}`);
  }

  return {
    article: processMeta({ ...articleUpdateResult.get, _id: id }),
    isCreated,
  };
}

export default {
  description: 'Create a reply request for the given article',
  type: new GraphQLObjectType({
    name: 'CreateReplyRequestResult',
    fields: {
      replyRequestCount: {
        type: GraphQLInt,
        description:
          'Reply request count for the given article after creating the reply request',
      },
      status: {
        type: new GraphQLEnumType({
          name: 'CreateReplyRequstResultStatus',
          values: {
            SUCCESS: {
              value: 'SUCCESS',
              description: 'Successfully inserted a new reply request',
            },
            DUPLICATE: {
              value: 'DUPLICATE',
              description:
                'The user has already requested reply for this article',
            },
          },
        }),
      },
    },
  }),
  args: {
    articleId: { type: new GraphQLNonNull(GraphQLString) },
    reason: {
      type: GraphQLString,
      description: 'The reason why the user want to submit this article',
    },
  },
  async resolve(rootValue, { articleId, reason }, { appId, userId }) {
    const { article, isCreated } = await createReplyRequest({
      articleId,
      appId,
      userId,
      reason,
    });
    return {
      replyRequestCount: article.replyRequestCount,
      status: isCreated ? 'SUCCESS' : 'DUPLICATE',
    };
  },
};
