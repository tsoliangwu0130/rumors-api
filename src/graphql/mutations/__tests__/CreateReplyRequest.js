import gql from 'util/GraphQL';
import { loadFixtures, unloadFixtures, resetFrom } from 'util/fixtures';
import client from 'util/client';
import MockDate from 'mockdate';
import { getReplyRequestId } from '../CreateReplyRequest';
import fixtures from '../__fixtures__/CreateReplyRequest';

describe('CreateReplyRequest', () => {
  beforeAll(() => loadFixtures(fixtures));

  it('attaches a reply request to an article', async () => {
    MockDate.set(1485593157011);
    const articleId = 'createReplyRequestTest1';
    const userId = 'test';
    const appId = 'test';

    const { data, errors } = await gql`
      mutation($articleId: String!) {
        CreateReplyRequest(articleId: $articleId, reason: "気になります") {
          replyRequestCount
          status
        }
      }
    `(
      {
        articleId,
      },
      { userId, appId }
    );
    MockDate.reset();

    const id = getReplyRequestId({ articleId, userId, appId });
    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot();

    const request = await client.get({
      index: 'replyrequests',
      type: 'doc',
      id,
    });
    expect(request._source).toMatchSnapshot();

    const article = await client.get({
      index: 'articles',
      type: 'doc',
      id: articleId,
    });
    expect(article._source).toMatchSnapshot();

    // Cleanup
    await client.delete({ index: 'replyrequests', type: 'doc', id });
    await resetFrom(fixtures, `/articles/doc/${articleId}`);
  });

  it('cannot attach a reply request to an article twice', async () => {
    MockDate.set(1485593157011);
    const articleId = 'createReplyRequestTest1';
    const userId = 'test';
    const appId = 'test';

    await gql`
      mutation($articleId: String!) {
        CreateReplyRequest(articleId: $articleId) {
          replyRequestCount
        }
      }
    `({ articleId }, { userId, appId });

    MockDate.set(1485593257011);

    const { data, errors } = await gql`
      mutation($articleId: String!) {
        CreateReplyRequest(articleId: $articleId) {
          replyRequestCount
          status
        }
      }
    `({ articleId }, { userId, appId });

    MockDate.reset();

    const id = getReplyRequestId({ articleId, userId, appId });
    expect(errors).toBeUndefined();
    expect(data).toMatchSnapshot();

    const conn = await client.get({
      index: 'replyrequests',
      type: 'doc',
      id,
    });
    expect(conn._source).toMatchSnapshot();

    const article = await client.get({
      index: 'articles',
      type: 'doc',
      id: articleId,
    });
    expect(article._source).toMatchSnapshot();

    // Cleanup
    await client.delete({ index: 'replyrequests', type: 'doc', id });
    await resetFrom(fixtures, `/articles/doc/${articleId}`);
  });

  afterAll(() => unloadFixtures(fixtures));
});
