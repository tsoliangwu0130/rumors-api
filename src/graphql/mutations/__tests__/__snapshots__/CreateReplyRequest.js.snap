// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateReplyRequest attaches a reply request to an article 1`] = `
Object {
  "CreateReplyRequest": Object {
    "replyRequestCount": 2,
    "status": "SUCCESS",
  },
}
`;

exports[`CreateReplyRequest attaches a reply request to an article 2`] = `
Object {
  "appId": "test",
  "articleId": "createReplyRequestTest1",
  "createdAt": "2017-01-28T08:45:57.011Z",
  "feedbacks": Array [],
  "negativeFeedbackCount": 0,
  "positiveFeedbackCount": 0,
  "reason": "気になります",
  "updatedAt": "2017-01-28T08:45:57.011Z",
  "userId": "test",
}
`;

exports[`CreateReplyRequest attaches a reply request to an article 3`] = `
Object {
  "lastRequestedAt": "2017-01-28T08:45:57.011Z",
  "replyRequestCount": 2,
  "text": "foofoo",
}
`;

exports[`CreateReplyRequest cannot attach a reply request to an article twice 1`] = `
Object {
  "CreateReplyRequest": Object {
    "replyRequestCount": 2,
    "status": "DUPLICATE",
  },
}
`;

exports[`CreateReplyRequest cannot attach a reply request to an article twice 2`] = `
Object {
  "appId": "test",
  "articleId": "createReplyRequestTest1",
  "createdAt": "2017-01-28T08:45:57.011Z",
  "feedbacks": Array [],
  "negativeFeedbackCount": 0,
  "positiveFeedbackCount": 0,
  "reason": "",
  "updatedAt": "2017-01-28T08:47:37.011Z",
  "userId": "test",
}
`;

exports[`CreateReplyRequest cannot attach a reply request to an article twice 3`] = `
Object {
  "lastRequestedAt": "2017-01-28T08:47:37.011Z",
  "replyRequestCount": 2,
  "text": "foofoo",
}
`;
